File: Calculators/Source/factorialGeoseriesCalc.cpp
/* FACTORIAL AND GEOSERIES CALCULATOR
 * William Wadsworth
 * CSC1710
 * Created: 10.15.2020
 * Doctored: 10.15.2023
 * ~/csc1710/lab9/wadsworthlab9.cpp
 * 
 * The binary was last compiled on 5.24.2024.
 * 
 * Usage:
 * To compile: g++ factorialGeoseriesCalc.cpp -Wall -o <exe name>
 * To run: ./<exe name>
*/

#include <iostream>
#include <sstream>
using namespace std;

// function prototpyes
long int factorial(int);
long int dfactorial(int);
double geoseries(double, int, double);

int main () 
{
   // declare variables
   long int endPoint;
   //int x, y = 1, prod = 1;


   // title
   cout << "Factorial and Geometric series calculations" << endl << " " << endl;

   // ask user preference
   string inp;
   cout << "Enter 'factorial', 'dfactorial', or 'geometric': ";
   cin >> inp;
   // check inp
   while(inp != "factorial" && inp != "dfactorial" && inp != "geometric") {
      cout << "Please enter 'factorial', 'dfactorial', or 'geometric': ";
      cin >> inp;
   }


   // if the user wants factorial calculation
   if (inp == "factorial") {
      // parameters for input, store in variable
      cout << "Integer must be between 0 and 1,000" << endl << " " << endl;
      cout << "Enter an integer to the nearest whole for factorial calculation: ";
      cin >> endPoint;

      // input validation
      while (endPoint > 1000 || endPoint < 0) {
         cout << "Not valid, integer must be between 0 and 1,000: ";
         cin >> endPoint;
      }

      // output calculation
      cout << endPoint << "! = " << factorial(endPoint) << endl;
   }



   // if the user wants double factorial calculation
   else if (inp == "dfactorial") {
      cout << "Integer must be between 0 and 1000" << endl << " " << endl;
      cout << "Enter an integer to the nearest whole for double factorial calculation: ";
      cin >> endPoint;

      // input validation, integer must be odd
      while (endPoint % 2 == 0 || endPoint > 1000 || endPoint < 0) {
         cout << "Not valid, integer must be odd and between 0 and 1,000: ";
         cin >> endPoint;
      }

      // output calculation
      cout << endPoint << "!! = " << dfactorial(endPoint) << endl;
   }



   else {
      // declare variables
      double a, r = 0.5, t;
      //double sum = 0;
      //int count;

      // title, ask for input
      cout << "Sum of Geometric series" << endl << " " << endl;
      cout << "What is your first term: ";
      cin >> a;

      cout << "How many terms would you like to take the sum of: ";
      cin >> t;

      cout << "Enter your common ratio of choice (enter -1 for 0.5): ";
      cin >> r;
      if(r == -1)
         r = 0.5;

      // display output
      cout << "Sum of " << t << " terms, a = " << a << ", r = " << r << ", is " << geoseries(a, t, r) << endl;
   }

   return 0;
}

// factorial function
long int factorial (int endPoint)
{
   //long int epoint;
   long int prod = 1;

   for (int x=1; x <= endPoint; x++) {
      prod *= x;
      //cout << prod << " ";
   }

   return prod;
}

// double factorial function
long int dfactorial (int endPoint)
{
   //long int epoint;
   int x = 1, prod = 1;

   do
   {
      prod *= x;
      x += 2;
   }
   while (x <= endPoint);

   return prod;
}

// geometric series function
double geoseries (double a, int t, double r/*=0.5*/)
{
   double sum = 0;

   for (int i=0; i < t; i++) {
      sum += a;
      a *= r;
   }

   return sum;
} 

File: Calculators/Source/factorialGeoseriesCalc.py
# FACTORIAL AND GEOSERIES CALCULATOR -- V.PY
# William Wadsworth
# CSC1710
# Created: 10.15.2020
# Doctored: 10.15.2023
# 
# Python-ized: 3.19.2024
# Updated 8.18.2024: PEP 8 Compliance
#  
# This program performs a factorial or geoseries calculation based on a number given from input.
# 
# Usage: python3 factorialGeoseriesCalc.py


# --- FUNCTIONS -------------------------------------------------------------------------
# --- FACTORIAL ---------------------------------
# pre-condition: endpoint must be a non-negative integer (0 <= endpoint <= 1000)
# post-condition: returns the factorial of endpoint
def factorial(endpoint: int) -> int:
    """Calculate the factorial of a given number."""
    
    prod: int = 1
    
    for x in range(1,endpoint+1):
        prod *= x
    
    return prod


# --- DOUBLE FACTORIAL --------------------------
# pre-condition: endpoint must be an odd non-negative integer (0 <= endpoint <= 1000)
# post-condition: returns the double factorial of endpoint
def double_factorial(endpoint: int) -> int:
    """Calculate the double factorial of a given odd number."""
    
    x: int = 1
    prod: int = 1
    
    while x <= endpoint:
        prod *= x
        x += 2
    
    return prod


# --- GEOSERIES ---------------------------------
# pre-condition: t must be a non-negative integer, a and r must be real numbers
# post-condition: returns the sum of the first t terms of the geometric series with first term a
#                 and common ratio r
def geoseries(a: float, t: int, r: float =0.5) -> float:
    """Calculate the sum of a geometric series."""
    
    sum: float = 0.0
    
    for _ in range(t):
        sum += a
        a *= r
        
    return sum


def main():
    # --- INTRODUCTION --------------------------
    print("Factorial and Geometric series calculations")

    # --- OPERATION SELECTION -------------------
    inp = input("Select operation 'factorial', 'dfactorial', or 'geometric': ")
    # input validation
    while inp not in ("factorial", "dfactorial", "geometric"):
        inp = input("error: invalid operation: ")

    # --- FACTORIAL -----------------------------
    if inp == "factorial":
        # parameters for input, store in variable
        print("\nInteger must be between 0 and 1,000\n")
        
        endpoint = int(input("Enter an integer to the nearest whole for factorial calculation: "))
        # input validation
        while endpoint < 0 or endpoint > 1000:
            endpoint = int(input("Not valid, integer must be between 0 and 1,000: "))
        
        # ouput calculation
        print(f"{endpoint}! = {factorial(endpoint)}")

    # --- DOUBLE FACTORIAL ----------------------
    elif inp == "dfactorial":
        # parameters for input, store in variable
        print("\nInteger must be between 0 and 1,000\n")
        
        endpoint = int(input("Enter an odd integer to the nearest whole for double factorial calculation: "))
        # input validation
        while endpoint % 2 == 0 or endpoint < 0 or endpoint > 1000:
            endpoint = int(input("Not valid, integer must be odd and between 0 and 1,000: "))
        
        # output calculation
        print(f"{endpoint}!! = {double_factorial(endpoint)}")

    # --- GEOSERIES -----------------------------
    # assume geoseries
    else:
        print("Sum of Geometric series")
        
        a = float(input("What is your first term: "))    
        t = int(input("How many terms would you like to take the sum of: "))
        # this is to avoid converting an empty string to a float
        try:
            r = float(input("Enter your common ratio of choice (leave blank for 0.5): "))
        except ValueError:
            r = 0.5
        
        print(f"Sum of {t} terms, a = {a}, r = {r}, is {geoseries(a,t,r)}")


if __name__ == "__main__":
    main()